conan/tools/sbom/cyclonedx.py

```diff
@@ -99,3 +99,103 @@ def _calculate_licenses(component):
+        add_build (bool, optional, default=False): Include build dependencies.
+        add_tests (bool, optional, default=False): Include test dependencies.
+
+    Returns:
+        The generated CycloneDX 1.6 document as a string.
+
+    Example usage:
+    ```
+    cyclonedx(conanfile, name="custom_name", add_build=True, add_test=True, **kwargs)
+    ```
+
+    """
+    import uuid
+    import time
+    from datetime import datetime, timezone
+    graph = conanfile.subgraph
+
+    has_special_root_node = not (getattr(graph.root.ref, "name", False) and getattr(graph.root.ref, "version", False) and getattr(graph.root.ref, "revision", False))
+    special_id = str(uuid.uuid4())
+
+    name_default = getattr(graph.root.ref, "name", False) or "conan-sbom"
+    name_default += f"/{graph.root.ref.version}" if bool(getattr(graph.root.ref, "version", False)) else ""
+    nodes = [node for node in graph.nodes if (node.context == "host" or add_build) and (not node.test or add_tests)]
+    if has_special_root_node:
+        nodes = nodes[1:]
+
+    dependencies = []
+    if has_special_root_node:
+        deps = {"ref": special_id,
+                "dependsOn": [f"pkg:conan/{d.dst.name}@{d.dst.ref.version}?rref={d.dst.ref.revision}"
+                              for d in graph.root.dependencies]}
+        dependencies.append(deps)
+    for c in nodes:
+        deps = {"ref": f"pkg:conan/{c.name}@{c.ref.version}?rref={c.ref.revision}"}
+        dep = [d for d in c.dependencies if (d.dst.context == "host" or add_build) and (not d.dst.test or add_tests)]
+
+        depends_on = [f"pkg:conan/{d.dst.name}@{d.dst.ref.version}?rref={d.dst.ref.revision}" for d in dep]
+        if depends_on:
+            deps["dependsOn"] = depends_on
+        dependencies.append(deps)
+
+    def _calculate_licenses(component):
+        if isinstance(component.conanfile.license, str): # Just one license
+            return [{"license": {
+                        "id": component.conanfile.license
+                    }}]
+        return [{"license": {
+                    "id": l
+                }} for l in component.conanfile.license]
+
+    sbom_cyclonedx_1_6 = {
+        **({"components": [{
+            **({"authors": [node.conanfile.author]} if node.conanfile.author else {}),
+            "bom-ref": special_id if has_special_root_node else f"pkg:conan/{node.name}@{node.ref.version}?rref={node.ref.revision}",
+            "description": node.conanfile.description,
+            **({"externalReferences": [{
+                "type": "website",
+                "url": node.conanfile.homepage
+            }]} if node.conanfile.homepage else {}),
+            **({"licenses": _calculate_licenses(node)} if node.conanfile.license else {}),
+            "name": node.name,
+            "purl": f"pkg:conan/{node.name}@{node.ref.version}",
+            "type": "application" if conanfile.package_type == "application" else "library",
```

Comment:
I think this is not right? Shouldn't it use `node.conanfile.package_type` ?

```suggestion
            "type": "application" if node.conanfile.package_type == "application" else "library",
```

Also, why is this hardcoded to `library` for 1.4? I think that you can also set it to `library` or `application` for 1.4 https://cyclonedx.org/docs/1.4/json/#metadata_component_type